File: java/src/main/java/com/fish/list/Q24.java
package com.fish.list;

/**
 * @author: liuqi
 * @date: 2024/3/6
 */
public class Q24 {
    /**
     * 递归解法
     */
    class Solution {
        public ListNode swapPairs(ListNode head) {
            if (head == null || head.next == null) {
                return head;
            }
            ListNode newHead = head.next;
            head.next = swapPairs(newHead.next);
            newHead.next = head;
            return newHead;
        }
    }


    /**
     * 迭代解法
     */
    class Solution1 {
        public ListNode swapPairs(ListNode head) {
            if (head == null || head.next == null) {
                return head;
            }
            ListNode preHead = new ListNode(-1, head);
            ListNode ret = head.next;
            while (head != null && head.next != null) {
                ListNode temp = head.next;
                head.next = temp.next;
                temp.next = head;
                preHead.next = temp;
                preHead = head;
                head = head.next;
            }
            return ret;
        }
    }
}

File: python/Q24.py
# Definition for singly-linked list.
class ListNode(object):
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution(object):
    def swapPairs(self, head):
        if not head or not head.next:
            return head
        pre = head
        post = head.next
        preHead = ListNode(-1, head)
        newHead = head.next
        while pre and post:
            # 进行 swap
            pre.next = post.next
            post.next = pre
            preHead.next = post
            # 更新指针
            if not pre.next or not pre.next.next:
                return newHead
            preHead = pre
            pre = pre.next
            post = pre.next
        return newHead
